File: tests/test_insights_layout.py
import re
import unittest
from pathlib import Path


class InsightsLayoutTest(unittest.TestCase):
    """Validate Insights card layout and centralized bullet styling."""

    @classmethod
    def setUpClass(cls):
        cls.root = Path(__file__).resolve().parent.parent
        cls.insights_html = (cls.root / "about/insights/index.html").read_text(encoding="utf-8")
        cls.site_css = (cls.root / "styles/site.css").read_text(encoding="utf-8")
        cls.insights_css = (cls.root / "styles/insights.css").read_text(encoding="utf-8")
        cls.services_css = (cls.root / "styles/services.css").read_text(encoding="utf-8")
        cls.services_html = (cls.root / "services/index.html").read_text(encoding="utf-8")

    def test_insights_has_grid_and_cards(self):
        """Ensure Insights page uses card grid classes."""
        self.assertIn('class="insights-grid"', self.insights_html)
        self.assertGreaterEqual(self.insights_html.count('class="insight-card card"'), 3)

    def test_insight_point_lists_use_bullet_list_class(self):
        """Ensure each Insights points list opts into the shared bullet system."""
        matches = re.findall(r'<ul class="([^"]*insight-points[^"]*)">', self.insights_html)
        self.assertGreaterEqual(len(matches), 3)
        for classes in matches:
            self.assertIn("bullet-list", classes)

    def test_services_lists_use_shared_bullet_list(self):
        """Ensure Services list markup also uses the shared bullet utility class."""
        matches = re.findall(r'<ul class="([^"]*service-list[^"]*)">', self.services_html)
        self.assertGreaterEqual(len(matches), 1)
        for classes in matches:
            self.assertIn("bullet-list", classes)

    def test_bullet_system_is_defined_in_site_css(self):
        """Ensure the shared bullet system and icon source are defined centrally."""
        self.assertIn(".bullet-list", self.site_css)
        self.assertIn(".bullet-list li::before", self.site_css)
        self.assertIn('background-image: url("/assets/icons/bullet.png");', self.site_css)

    def test_page_css_does_not_override_bullet_pseudo_elements(self):
        """Ensure page CSS files do not define competing bullet pseudo-element rules."""
        self.assertNotRegex(self.insights_css, r"insight-points\s+li::before")
        self.assertNotRegex(self.insights_css, r"bullet-list\s+li::before")
        self.assertNotRegex(self.services_css, r"service-list\s+li::before")
        self.assertNotRegex(self.services_css, r"bullet-list\s+li::before")


if __name__ == "__main__":
    unittest.main()

File: tests/test-insights-layout.js
#!/usr/bin/env node
/**
 * Test: Insights structure, styling behavior, and shared orb bullet usage.
 */
const fs = require('fs');
const path = require('path');

const root = path.resolve(__dirname, '..');
const insightsHtml = fs.readFileSync(path.join(root, 'insights/index.html'), 'utf8');
const insightsCss = fs.readFileSync(path.join(root, 'styles/insights.css'), 'utf8');
const siteCss = fs.readFileSync(path.join(root, 'styles/site.css'), 'utf8');
const insightsScript = fs.readFileSync(path.join(root, 'scripts/insights-briefs.js'), 'utf8');

let failures = 0;

// Validate card markup and toggle wiring.
const insightCards = [...insightsHtml.matchAll(/<section id="([a-z0-9-]+)" class="insight-card">([\s\S]*?)<\/section>/g)];
if (insightCards.length < 3) {
  failures += 1;
  console.error('FAIL: Insights page should contain at least 3 insight-card sections.');
}

const seenSlugs = new Set();
for (const card of insightCards) {
  const slug = card[1];
  const cardHtml = card[2];
  const titleMatch = cardHtml.match(/<h2>([^<]+)<\/h2>/);

  if (!slug || seenSlugs.has(slug)) {
    failures += 1;
    console.error(`FAIL: duplicate or missing card slug: ${slug}`);
  }
  seenSlugs.add(slug);

  if (!titleMatch) {
    failures += 1;
    console.error(`FAIL: card ${slug} is missing an h2 title.`);
  }

  if (!/<button class="insight-toggle"[^>]*aria-expanded="false"/i.test(cardHtml)) {
    failures += 1;
    console.error(`FAIL: card ${slug} is missing the collapsed insight-toggle button.`);
  }

  if (!/<ul class="service-list">/i.test(cardHtml)) {
    failures += 1;
    console.error(`FAIL: card ${slug} is missing shared service-list bullets.`);
  }

  if (!/<div class="insight-expanded">/i.test(cardHtml)) {
    failures += 1;
    console.error(`FAIL: card ${slug} is missing insight-expanded content container.`);
  }
}

if (!insightsHtml.includes('src="../scripts/insights-briefs.js"')) {
  failures += 1;
  console.error('FAIL: Insights page must include insights-briefs.js from ../scripts/.');
}

// Validate CSS behavior for collapse/expand and hover lift.
if (!/\.insight-expanded\s*\{[^}]*display:\s*none;/s.test(insightsCss)) {
  failures += 1;
  console.error('FAIL: insight-expanded must be hidden by default.');
}
if (!/\.insight-card\.expanded\s+\.insight-expanded\s*\{[^}]*display:\s*block;/s.test(insightsCss)) {
  failures += 1;
  console.error('FAIL: expanded insight content rule is missing.');
}
if (!/\.insight-card\s*\{[^}]*transition:\s*transform 150ms ease;/s.test(insightsCss)) {
  failures += 1;
  console.error('FAIL: insight-card transition must use transform 150ms ease.');
}
if (!/\.insight-card:hover\s*\{[^}]*transform:\s*translateY\(-2px\);/s.test(insightsCss)) {
  failures += 1;
  console.error('FAIL: insight-card hover lift must be translateY(-2px).');
}

// Validate shared orb bullet implementation in global CSS.
if (!/\.service-list li::before\s*\{[^}]*width:\s*10px;[^}]*height:\s*10px;[^}]*margin-right:\s*12px;[^}]*background-image:\s*url\("\/assets\/icons\/bullet\.png"\);/s.test(siteCss)) {
  failures += 1;
  console.error('FAIL: shared orb bullet spec is not correctly defined in site.css.');
}

// Validate GA tracking contract in script.
if (!insightsScript.includes("gtag('event', 'insight_expand'")) {
  failures += 1;
  console.error('FAIL: insight_expand GA event call is missing.');
}
if (!/if \(isExpanded && typeof gtag === 'function'\)/.test(insightsScript)) {
  failures += 1;
  console.error('FAIL: GA event must fire only on expand and guard missing gtag.');
}

if (failures > 0) {
  process.exit(1);
}

console.log('PASS: insights structure, behavior, and orb bullet checks passed.');
